src/PlusDataCollection/PointGrey/vtkPlusSpinnakerVideoSource.cxx

```diff
@@ -0,0 +1,807 @@
+PlusStatus vtkPlusSpinnakerVideoSource::CheckCameraParameterValidity()
+{
+  PlusStatus retVal = PLUS_SUCCESS;
+
+  // exposure
+  if (this->ExposureMode == EXPOSURE_TIMED && this->ExposureMicroSec == FLAG_EXPOSURE_MICROSEC)
+  {
+    LOG_ERROR("Failed to configure SpinnakerVideoSource: ExposureMode is set to \"TIMED\", but no ExposureMicroSec attribute is provided.");
+    retVal = PLUS_FAIL;
+  }
+
+  // gain
+  if (this->GainMode == GAIN_MANUAL && this->GainDB == FLAG_GAIN_DB)
+  {
+    LOG_ERROR("Failed to configure SpinnakerVideoSource: GainMode is set to \"MANUAL\", but no GainDB attribute is provided.");
+    retVal = PLUS_FAIL;
+  }
+
+  // white balance
+  if (this->WhiteBalanceMode == WB_MANUAL && this->WhiteBalanceRed == FLAG_WHITE_BALANCE)
+  {
+    LOG_ERROR("Failed to configure SpinnakerVideoSource: WhiteBalanceMode is set to \"MANUAL\", but no WhiteBalanceRed attribute is provided.");
+    retVal = PLUS_FAIL;
+  }
+  if (this->WhiteBalanceMode == WB_MANUAL && this->WhiteBalanceBlue == FLAG_WHITE_BALANCE)
+  {
+    LOG_ERROR("Failed to configure SpinnakerVideoSource: WhiteBalanceMode is set to \"MANUAL\", but no WhiteBalanceBlue attribute is provided.");
+    retVal = PLUS_FAIL;
+  }
+
+  // sharpening
+  if (this->SharpeningMode == SHARPENING_MANUAL && this->SharpeningAmount == FLAG_SHARPENING)
+  {
+    LOG_ERROR("Failed to configure SpinnakerVideoSource: SharpeningMode is set to \"MANUAL\", but no SharpeningAmount attribute is provided.");
+    retVal = PLUS_FAIL;
+  }
+
+  return retVal;
+}
+
+//-----------------------------------------------------------------------------
+PlusStatus vtkPlusSpinnakerVideoSource::ReadConfiguration(vtkXMLDataElement* rootConfigElement)
+{
+  LOG_TRACE("vtkPlusSpinnakerVideoSource::ReadConfiguration(vtkXMLDataElement* rootConfigElement)");
+
+  XML_FIND_DEVICE_ELEMENT_REQUIRED_FOR_READING(deviceConfig, rootConfigElement);
+
+  // camera params
+  XML_READ_SCALAR_ATTRIBUTE_REQUIRED(int, CameraNumber, deviceConfig);
+  XML_READ_ENUM3_ATTRIBUTE_OPTIONAL(PixelEncoding, deviceConfig,
+    "RGB24", RGB24,
+    "BGR24", BGR24,
+    "MONO8", MONO8);
+
+  // frame size
+  int requestedFrameSize[2] = { static_cast<int>(DEFAULT_FRAME_SIZE[0]), static_cast<int>(DEFAULT_FRAME_SIZE[1]) };
+  if (deviceConfig->GetVectorAttribute("FrameSize", 2, requestedFrameSize))
+  {
+    if (requestedFrameSize[0] < 0 || requestedFrameSize[1] < 0)
+    {
+      LOG_ERROR("Negative frame size defined in config file. Please define a positive frame size.");
+      return PLUS_FAIL;
+    }
+    this->FrameSize[0] = static_cast<unsigned int>(requestedFrameSize[0]);
+    this->FrameSize[1] = static_cast<unsigned int>(requestedFrameSize[1]);
+    this->FrameSize[2] = 1;
+  }
+
+  // frame rate
+  XML_READ_SCALAR_ATTRIBUTE_OPTIONAL(int, FrameRate, deviceConfig);
+
+  // exposure
+  XML_READ_ENUM3_ATTRIBUTE_OPTIONAL(ExposureMode, deviceConfig,
+    "TIMED", EXPOSURE_TIMED,
```

Comment:
If you re-use these strings with PrintConfiguration, I would move them to static/anonymous vars

Reply:
As a std::map?

Reply:
Nah, just a bunch of static or anonymous namespaced std::string

Reply:
Or const char* if they're mostly used in VTK functions that don't like std::string